### ISSUE TITLE
Support custom keychain on macOS to support non-root daemon processes

### ISSUE THREAD

**Issue by EeroKurimo** [aws/aws-iot-device-sdk-cpp-v2#163]
**Is your feature request related to a problem? Please describe.**

While investigating the issue https://github.com/aws/aws-iot-device-sdk-cpp-v2/issues/162, we realized that while it's somewhat trivial to enable access to the default login keychain of an account to the IoT SDK, this is not the case when doing it in a daemon process. When launching an aws iot -enabled daemon process under the account `daemon` through `launchctl`, the initialization fails with the error `ERROR: AWS-pki-utils: static: error importing certificate with OSStatus -61` (i.e. `errSecWrPerm`) and at the same time the `log stream` command shows the error `[com.apple.securityd:atomicfile] create /Library/Keychains/System.keychain.sb-de8c63bf-RyxTn0: Permission denied`. 

**Describe the solution you'd like**

Would it be possible to support optional custom keychains as opposed to the current default login keychain only?

Currently in https://github.com/awslabs/aws-c-io/blob/ca96a75aecca05b8c25f0464a303b06f5a5a51af/source/darwin/darwin_pki_utils.c#L48 you open the default login keychain with `SecKeychainCopyDefault()`, but if `aws_import_public_and_private_keys_to_identity` could have an optional argument e.g. `import_keychain_path`, it could also support custom keychains for the daemon use case, with something like this in its place:

```c++
SecCertificateRef certificate_ref = NULL;
SecKeychainRef import_keychain = NULL;

OSStatus keychain_status;
CFStringRef import_password = CFSTR("");

if (!import_keychain_path) {
	keychain_status = SecKeychainCopyDefault(&import_keychain);
} else {
	keychain_status = SecKeychainCreate(
		import_keychain_path, 
		CFStringGetLength(import_password), 
		import_password, 
		false, 
		NULL, 
		&import_keychain);
	if (keychain_status == errSecDuplicateKeychain) {
		keychain_status = SecKeychainOpen(import_keychain_path, &import_keychain);
	}
}

if (keychain_status != errSecSuccess) {
	AWS_LOGF_ERROR(AWS_LS_IO_PKI, "static: error opening keychain with OSStatus %d", keychain_status);
	return AWS_OP_ERR;
}
```

Quickly trying it with a hardcoded keychain path seemed to fix our daemon use case.

Naturally the C++ layer would also need some way to optionally define the custom keychain path. Maybe by adding a few overloaded macOS-only `TlsContextOptions::InitClientWithMtls` functions that have an keychain path as an argument into https://github.com/awslabs/aws-crt-cpp/blob/master/source/io/TlsOptions.cpp? 

I'm also open to other options, as long as we could still have a non-root daemon processes connecting to aws iot.

**Describe alternatives you've considered**

The alternative for daemon processes would be to run them as `root` instead of a lower priviledge account, but from a security point of view requiring that just for authenticating through TLS to aws sounds dangeous.

**Comment by graebm** [comment]
I'm looking into this...

Could you work around the issue right now by calling [SecKeychainSetDefault()](https://developer.apple.com/documentation/security/1393097-seckeychainsetdefault?language=objc) during application startup? That way, when the SDK calls `SecKeychainCopyDefault()`, it will get the keychain of your choice.

No one on our team is an expert with Apple's security API. We're looking into if it's possible to **NOT** involve the keychain at all when creating an identity from a certificate and private-key...

**Comment by EeroKurimo** [comment]
I need to test it, but that sounds like a very promising workaround we can live with for now.

I'll also ask around for possible suggestions from colleagues who might more more macOS security API expertise than I do.

**Comment by graebm** [comment]
OK, looked into it more and I don't see any way to create an identity from a certificate and private-key without using the keychain.
### Investigation
Looking at the [Identities API](https://developer.apple.com/documentation/security/certificate_key_and_trust_services/identities?language=objc) I didn't see any way to do it. We can call [SecIdentityCreateWithCertificate()](https://developer.apple.com/documentation/security/certificate_key_and_trust_services/identities/creating_an_identity?language=objc) without passing a keychain, but then it just uses the default keychain anyway. We could create a "junk" keychain (like in your code sample above), but that junk keychain still needs to live on disk somewhere. [SecKeychainCreate()](https://developer.apple.com/documentation/security/1401214-seckeychaincreate?language=objc) requires a `const char *pathName`. I tried passing `NULL` and `""` and they were rejected.

This [Apple Developer Forums post](https://developer.apple.com/forums/thread/656952?answerId=627114022#627114022) (answered by an Apple Developer Support person) seems to agree with this:
> The problem here is forming the identity. It’s not possible to construct a `SecIdentity` object from an in-memory private key and its matching certificate. Your options are:
> * Add the private key to the keychain and then use `SecIdentityCreateWithCertificate` (this is a Mac-only routine, FYI).
> * Add both to the keychain and then get the identity using `SecItemCopyMatching`.
> * Give up on this approach and use a PKCS#12, which you can import directly.

This [Stack Overflow post](https://stackoverflow.com/questions/18284689/secure-transport-load-server-certificate-from-file) has the same question, and the one answer just advises them to use PKCS#12.

### Conclusion
If you can use PKCS#12, do that. The Apple API lets us create an identity from that without touching the keychain.

We're going to leave things as they are unless you can't work around this and need our APIs to take an `import_keychain_path` like in your code sample above. But we would probably use [SecKeychainOpen()](https://developer.apple.com/documentation/security/1396431-seckeychainopen?language=objc) instead of [SecKeychainCreate()](https://developer.apple.com/documentation/security/1401214-seckeychaincreate?language=objc). So the user would still need to have created the keychain before calling into our APIs

**Comment by EeroKurimo** [comment]
The stored cert+privkey is used by passing it to [`SSLSetCertificate()`](https://github.com/awslabs/aws-c-io/blob/03f6c7e858ba40da71450bb7ce090a17a27facb9/source/darwin/secure_transport_tls_channel_handler.c#L880), and its [documentation](https://developer.apple.com/documentation/security/1392400-sslsetcertificate?language=occ) insists that it gets a [`SecCertificateRef`](https://developer.apple.com/documentation/security/SecCertificateRef?language=occ), which is an "_object for a certificate that is stored in a keychain can be safely cast to a SecKeychainItemRef for manipulation as a keychain item_". This means it indeed looks like there's no apparent reasonable way of using the [Secure Transport API](https://developer.apple.com/documentation/security/secure_transport?language=occ) without keychains being involved. The newer [Network framework](https://developer.apple.com/documentation/network?language=occ) intending to replace it might be a different case, but that's available only from macOS 10.14 onwards, and we're hoping to support at least 10.13, so we're happier to stick with the Secure Transport for now.

One apparent problem with using `SecKeychainSetDefault()` is that it'd also affect all APIs that rely on login keychain for the process, like all networking unrelated to AWS IoT. However, this might not be a problem for us, as we'll be enforcing our own pinned certificates when connecting to our services outside of AWS IoT.

**Comment by EeroKurimo** [comment]
As an update for this case, we tried changing the default keychain of the daemon-account with `SecKeychainSetDefault`, but seemed to encounter some strange errors when testing with it. Using that API also seemed a bit hazardous for library usage, as loading the shared library messed up the default keychain of at least one developers testing it under his own account. We're not also sure can the default keychain be safely changed back to the default default keychain, as suggested by the documentation of `SecKeychainSetDefault`:

>You may call this function to change where a password or other keychain items are added, but since this is a user choice, you should set the default keychain back to the user specified keychain when you are done.

The alternative of using the PKCS#12 unfortunately doesn't sound like a good solution for us. Adding custom code generating such files or linking e.g. openssl to do that just for creating a container file for cert+key is getting a bit too close to overengineering just to work around a missing API to use a private keychain. It also looks like the PKCS#12 functions would themselves rely on using the default keychain to create Identity objects, as suggested by the Apple Documentation about [Importing an Identity](https://developer.apple.com/documentation/security/certificate_key_and_trust_services/identities/importing_an_identity?language=objc):

> **Note** 
> In macOS, when it succeeds, the SecPKCS12Import function call automatically stores the extracted identity in the default keychain in addition to returning a reference to it as one of the items in the array of dictionaries.

Due to this we're now temporarily using a locally patched version of the IoT SDK which supports an optional private keychain path argument, which uses `SecKeychainOpen(import_keychain_path, &import_keychain)` instead of `SecKeychainCopyDefault(&import_keychain)` when a customer keychain path is defined, but leaving the responsibility of creating the private keychain to whoever wants to use one.

I could make a pull request out of these changes to the aws-c-io and aws-crt-cpp repositories? We're also open for other suggestions.
